### PR TITLE
Logout functionality

### LINKED ISSUE #86: logout button
Eventually if we logged in on some computer we want to be able to log out as well -- i.e. if it is a shared computer. So, a button "log out" which invalidates the cookie would be great!

### PR DESCRIPTION
**Pull request by PizieDust**
closes #86 #85 

This PR adds buttons to close single sessions and also a logout button.

### CODE REVIEW

File: unikernel.ml [hunk 1]
```diff
@@ -661,31 +673,48 @@ struct
...
+    Store.update_user !store updated_user >>= function
+    | Ok store' ->
+        store := store';
+        redirect
+    | Error (`Msg err) ->
+        Logs.warn (fun m -> m "Storage error with %s" err);
+        Middleware.http_response reqd ~title:"Error" ~data:err
+          `Internal_server_error
+
+  let close_sessions ?(single_cookie_value = "") ?(logout = false) store reqd
```
Comment (hannesm) [inline unikernel.ml:691]:
```suggestion
  let close_sessions ?single_cookie_value ?(logout = false) store reqd
```

File: unikernel.ml [hunk 2]
```diff
@@ -661,31 +673,48 @@ struct
...
-                   (fun (cookie : User_model.cookie) ->
-                     not (String.equal cookie.name User_model.session_cookie))
-                   user.cookies
-            in
-            let updated_user =
-              User_model.update_user user ~cookies ~updated_at:now ()
+        | Some cookie ->
+            let filter, redirect =
+              match (String.equal single_cookie_value "", logout) with
+              | true, false ->
```
Comment (hannesm) [inline unikernel.ml:699]:
```suggestion
              match (single_cookie_value, logout) with
              | None, false ->
```

File: assets/main.js [hunk 3]
```diff
@@ -1,5 +1,16 @@
...
 	AOS.init();
+
+	const flashMessage = getCookie('flash_msg');
+	if (flashMessage) {
+		if (flashMessage.startsWith("error:")) {
+			postAlert("bg-secondary-300", flashMessage);
+		} else {
+			postAlert("bg-primary-300", flashMessage);
+		}
+		document.cookie = "flash_msg=; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 UTC;";
```
[the remark this answers is not in the document]
Comment (PizieDust) [reply assets/main.js:11]: so currently we don't have a way to redirect to a new page and display a message or notification. This cookie is deleted immediately it is displayed to the user. The cookie carries the message.

File: unikernel.ml [hunk 4]
```diff
@@ -661,31 +673,48 @@ struct
...
+              | _, true ->
+                  ( (fun (c : User_model.cookie) ->
+                      not (String.equal c.value cookie.value)),
+                    Middleware.http_response reqd ~title:"OK"
+                      ~data:"Logout succesful" `OK )
+              | _ ->
+                  ( (fun (c : User_model.cookie) ->
+                      not (String.equal single_cookie_value c.value)),
+                    Middleware.redirect_to_page ~path:"/account"
+                      ~msg:"Closed session succesfully" reqd () )
```
Comment (hannesm) [inline unikernel.ml:715]:
```suggestion
              | Some to_logout_cookie_value, false ->
                  ( (fun (c : User_model.cookie) ->
                      not (String.equal to_logout_cookie_value c.value)),
                    Middleware.redirect_to_page ~path:"/account"
                      ~msg:"Closed session succesfully" reqd () )
```

File: unikernel.ml [hunk 5]
```diff
@@ -1324,9 +1362,34 @@ struct
...
+                extract_csrf_token reqd >>= function
+                | Ok (form_csrf, _) ->
+                    authenticate ~form_csrf store reqd
+                      (close_sessions ~logout:true store reqd)
+                | Error (`Msg msg) ->
+                    Middleware.http_response reqd ~title:"Error"
+                      ~data:(String.escaped msg) `Bad_request)
+        | path
+          when String.(
+                 length path >= 21 && sub path 0 23 = "/account/session/close/")
```
Comment (hannesm) [inline unikernel.ml:1376]:
```suggestion
                 length path >= 23 && sub path 0 23 = "/account/session/close/")
```

File: unikernel.ml [hunk 6]
```diff
@@ -661,31 +673,48 @@ struct
...
+    Store.update_user !store updated_user >>= function
+    | Ok store' ->
+        store := store';
+        redirect
+    | Error (`Msg err) ->
+        Logs.warn (fun m -> m "Storage error with %s" err);
+        Middleware.http_response reqd ~title:"Error" ~data:err
+          `Internal_server_error
+
+  let close_sessions ?single_cookie_value ?(logout = false) store reqd
```
Comment (hannesm) [inline unikernel.ml:691]:
```suggestion
  let close_sessions ?to_close_cookie ?(logout = false) store reqd
```

File: unikernel.ml [hunk 7]
```diff
@@ -661,31 +673,48 @@ struct
...
-              :: List.filter
-                   (fun (cookie : User_model.cookie) ->
-                     not (String.equal cookie.name User_model.session_cookie))
-                   user.cookies
-            in
-            let updated_user =
-              User_model.update_user user ~cookies ~updated_at:now ()
+        | Some cookie ->
+            let filter, redirect =
+              match (single_cookie_value, logout) with
```
Comment (hannesm) [inline unikernel.ml:698]:
```suggestion
              match (to_logout_cookie, logout) with
```

File: unikernel.ml [hunk 8]
```diff
@@ -1324,9 +1362,34 @@ struct
...
+                 length path >= 23 && sub path 0 23 = "/account/session/close/")
+          ->
+            check_meth `GET (fun () ->
+                match
+                  String.split_on_char '/'
+                    (String.sub path 23 (String.length path - 23))
+                with
+                | [ single_cookie_value; form_csrf ] ->
+                    authenticate ~form_csrf store reqd
+                      (close_sessions ~single_cookie_value store reqd)
```
Comment (hannesm) [inline unikernel.ml:1385]:
```suggestion
                      (close_sessions ~to_logout_cookie store reqd)
```

File: unikernel.ml [hunk 9]
```diff
@@ -1324,9 +1362,34 @@ struct
...
+        | path
+          when String.(
+                 length path >= 23 && sub path 0 23 = "/account/session/close/")
+          ->
+            check_meth `GET (fun () ->
+                match
+                  String.split_on_char '/'
+                    (String.sub path 23 (String.length path - 23))
+                with
+                | [ single_cookie_value; form_csrf ] ->
```
Comment (hannesm) [inline unikernel.ml:1383]:
```suggestion
                | [ to_logout_cookie; form_csrf ] ->
```

File: user_account.ml [hunk 10]
```diff
@@ -306,6 +305,27 @@ let user_account_layout ~csrf (user : User_model.user) ~active_cookie_value
...
+                                             ^ cookie.value ^ "/" ^ csrf);
+                                            a_class
+                                              [
+                                                "hover:text-secondary-800 \
+                                                 text-secondary-500 \
+                                                 transition-colors \
+                                                 cursor-pointer";
+                                              ];
+                                          ]
+                                        [ txt "Logout this session" ]
```
Comment (hannesm) [inline user_account.ml:327]:
```suggestion
                                        [ txt "Close session" ]
```

**Comment by hannesm** [comment]
Looks fine to me, only a few minor remarks above.

**Comment by PizieDust** [comment]
Thank you for the review @hannesm

**Comment by hannesm** [comment]
Thank you for the development :)

**Pull request merged**